compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -446,12 +447,195 @@ impl<'a> Parser<'a> {
             || self.token.kind == token::CloseDelim(Delimiter::Parenthesis)
                 && self.look_ahead(1, Token::is_range_separator);
 
+        let span = expr.span;
+
         Some((
-            self.dcx().emit_err(UnexpectedExpressionInPattern { span: expr.span, is_bound }),
-            expr.span,
+            self.dcx().stash_err(
+                span,
+                StashKey::ExprInPat,
+                UnexpectedExpressionInPattern { span, is_bound },
+            ),
+            span,
         ))
     }
 
+    /// Called by [`Parser::parse_stmt_without_recovery`], used to add statement-aware subdiagnostics to the errors stashed
+    /// by [`Parser::maybe_recover_trailing_expr`].
+    pub(super) fn maybe_emit_stashed_expr_in_pats(&mut self, stmt: &Stmt) {
+        if self.dcx().has_errors().is_none() {
+            // No need to walk the statement if there's no stashed errors.
+            return;
+        }
+
+        struct PatVisitor<'a> {
+            /// `self`
+            parser: &'a Parser<'a>,
+            /// The freshly-parsed statement.
+            stmt: &'a Stmt,
+            /// The current match arm (for arm guard suggestions).
+            arm: Option<&'a Arm>,
+            /// The current struct field (for variable name suggestions).
+            field: Option<&'a PatField>,
+        }
+
+        impl<'a> PatVisitor<'a> {
+            /// Looks for stashed [`StashKey::ExprInPat`] errors in `stash_span`, and emit them with suggestions.
+            /// `stash_span` is contained in `expr_span`, the latter being larger in borrow patterns;
+            /// ```txt
+            /// &mut x.y
+            /// -----^^^ `stash_span`
+            /// |
+            /// `expr_span`
+            /// ```
+            /// `is_range_bound` is used to exclude arm guard suggestions in range pattern bounds.
+            fn emit_now(&self, stash_span: Span, expr_span: Span, is_range_bound: bool) {
+                self.parser.dcx().try_steal_modify_and_emit_err(
+                    stash_span,
+                    StashKey::ExprInPat,
+                    |err| {
+                        let sm = self.parser.psess.source_map();
+                        let stmt = self.stmt;
+                        let line_lo = sm.span_extend_to_line(stmt.span).shrink_to_lo();
+                        let indentation = sm.indentation_before(stmt.span).unwrap_or_default();
+                        let expr = self.parser.span_to_snippet(expr_span).unwrap();
```

Comment:
Unwrapping `span_to_snippet` is pretty fishy but I can't think of a pathological example rn. Seems okay to keep for now.

Reply:
Actually, this can indeed ICE (similar to #128442, #128445 which I found after seeing this unwrap):

```rs
// a.rs
#[macro_export]
macro_rules! make {
    () => { match 0 { 0 + 0 => {} } }
}
```

```rs
// b.rs
fn main() { a::make!(); }
```

1. Compile `a`: `rustc +TOOLCHAIN a.rs --crate-type=lib`
2. Remove or move `a.rs`
3. Compile `b`: `rustc +TOOLCHAIN b.rs -L. --extern a --edition=2021`

---

Could you instead return early if `span_to_snippet` returns `None` and add a FIXME that `UnexpectedExpressionInPatternSugg::{Create,Update}Guard` don't need `span_to_snippet` / `expr : String` (as explained in another review comment of mine) meaning we don't need to return early for *them* (for `…Sugg::Const` it'll still be necessary tho).